lib/vector-buffers/src/topology/channel/receiver.rs

```diff
@@ -58,10 +59,20 @@ where
         match self {
             ReceiverAdapter::InMemory(rx) => rx.next().await,
             ReceiverAdapter::DiskV1(reader) => reader.next().await,
-            ReceiverAdapter::DiskV2(reader) => reader
-                .next()
-                .await
-                .expect("reader encountered unrecoverable error"),
+            ReceiverAdapter::DiskV2(reader) => loop {
+                match reader.next().await {
+                    Ok(result) => break result,
+                    Err(e) => match e.as_recoverable_error() {
+                        Some(re) => {
+                            // If we've hit a recoverable error, we'll emit an event to indicate as much but we'll still
+                            // keep trying to read the next available record.
+                            emit(re);
+                            continue;
+                        }
+                        None => panic!("Reader encountered unrecoverable error: {:?}", e),
+                    },
+                }
+            },
```

Comment:
We do this here, instead of in `disk_v2::Reader` itself, to leave the logic in the buffer a little simpler. Similar to acknowledgements, we want the control of "what do I do when a message has a negative ack?" to be pulled to the forefront, not hidden.